File: src/com/android/settings/network/MobileNetworkListFragment.java
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.settings.network;

import android.app.settings.SettingsEnums;
import android.content.Context;
import android.os.UserManager;
import android.provider.SearchIndexableResource;

import androidx.annotation.VisibleForTesting;

import com.android.settings.R;
import com.android.settings.Utils;
import com.android.settings.dashboard.DashboardFragment;
import com.android.settings.search.BaseSearchIndexProvider;
import com.android.settingslib.core.AbstractPreferenceController;
import com.android.settingslib.search.SearchIndexable;

import java.util.ArrayList;
import java.util.List;

@SearchIndexable(forTarget = SearchIndexable.ALL & ~SearchIndexable.ARC)
public class MobileNetworkListFragment extends DashboardFragment {
    private static final String LOG_TAG = "NetworkListFragment";

    static final String KEY_PREFERENCE_CATEGORY_SIM = "provider_model_sim_category";
    @VisibleForTesting
    static final String KEY_PREFERENCE_CATEGORY_DOWNLOADED_SIM =
            "provider_model_downloaded_sim_category";

    @Override
    protected int getPreferenceScreenResId() {
        return R.xml.network_provider_sims_list;
    }

    @Override
    protected String getLogTag() {
        return LOG_TAG;
    }

    @Override
    public int getMetricsCategory() {
        return SettingsEnums.MOBILE_NETWORK_LIST;
    }

    @Override
    protected List<AbstractPreferenceController> createPreferenceControllers(Context context) {
        final List<AbstractPreferenceController> controllers = new ArrayList<>();
        if (!SubscriptionUtil.isSimHardwareVisible(getContext())) {
            finish();
            return controllers;
        }

        NetworkProviderSimsCategoryController simCategoryPrefCtrl =
                new NetworkProviderSimsCategoryController(context, KEY_PREFERENCE_CATEGORY_SIM,
                        getSettingsLifecycle());
        controllers.add(simCategoryPrefCtrl);
        NetworkProviderDownloadedSimsCategoryController downloadedSimsCategoryCtrl =
                new NetworkProviderDownloadedSimsCategoryController(context,
                        KEY_PREFERENCE_CATEGORY_DOWNLOADED_SIM, getSettingsLifecycle());
        controllers.add(downloadedSimsCategoryCtrl);

        return controllers;
    }

    public static final BaseSearchIndexProvider SEARCH_INDEX_DATA_PROVIDER =
            new BaseSearchIndexProvider() {

                @Override
                public List<SearchIndexableResource> getXmlResourcesToIndex(Context context,
                        boolean enabled) {
                    final ArrayList<SearchIndexableResource> result = new ArrayList<>();
                    final SearchIndexableResource sir = new SearchIndexableResource(context);
                    sir.xmlResId = R.xml.network_provider_sims_list;
                    result.add(sir);
                    return result;
                }

                @Override
                protected boolean isPageSearchEnabled(Context context) {
                    return SubscriptionUtil.isSimHardwareVisible(context) &&
                            context.getSystemService(UserManager.class).isAdminUser();
                }
            };
}

File: src/com/android/settings/network/MobileNetworkListFragment.kt
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.settings.network

import android.app.settings.SettingsEnums
import android.content.Context
import android.os.Bundle
import android.provider.Settings
import android.view.View
import androidx.annotation.VisibleForTesting
import androidx.preference.Preference
import com.android.settings.R
import com.android.settings.SettingsPreferenceFragment
import com.android.settings.dashboard.DashboardFragment
import com.android.settings.flags.Flags
import com.android.settings.network.telephony.MobileNetworkUtils
import com.android.settings.search.BaseSearchIndexProvider
import com.android.settings.spa.SpaActivity.Companion.startSpaActivity
import com.android.settings.spa.network.NetworkCellularGroupProvider
import com.android.settingslib.search.SearchIndexable
import com.android.settingslib.spa.framework.util.collectLatestWithLifecycle
import com.android.settingslib.spaprivileged.framework.common.userManager
import com.android.settingslib.spaprivileged.settingsprovider.settingsGlobalBooleanFlow

@SearchIndexable(forTarget = SearchIndexable.ALL and SearchIndexable.ARC.inv())
class MobileNetworkListFragment : DashboardFragment() {
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        collectAirplaneModeAndFinishIfOn()
    }

    override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)

        if (Flags.isDualSimOnboardingEnabled()) {
            context?.startSpaActivity(NetworkCellularGroupProvider.name);
            finish()
        }
    }

    override fun onResume() {
        super.onResume()
        // Disable the animation of the preference list
        listView.itemAnimator = null

        findPreference<Preference>(KEY_ADD_SIM)!!.isVisible =
            MobileNetworkUtils.showEuiccSettings(context)
    }

    override fun getPreferenceScreenResId() = R.xml.network_provider_sims_list

    override fun getLogTag() = LOG_TAG

    override fun getMetricsCategory() = SettingsEnums.MOBILE_NETWORK_LIST

    companion object {
        private const val LOG_TAG = "NetworkListFragment"
        private const val KEY_ADD_SIM = "add_sim"

        @JvmStatic
        fun SettingsPreferenceFragment.collectAirplaneModeAndFinishIfOn() {
            requireContext().settingsGlobalBooleanFlow(Settings.Global.AIRPLANE_MODE_ON)
                .collectLatestWithLifecycle(viewLifecycleOwner) { isAirplaneModeOn ->
                    if (isAirplaneModeOn) {
                        finish()
                    }
                }
        }

        @JvmField
        val SEARCH_INDEX_DATA_PROVIDER = SearchIndexProvider()

        @VisibleForTesting
        class SearchIndexProvider : BaseSearchIndexProvider(R.xml.network_provider_sims_list) {
            public override fun isPageSearchEnabled(context: Context): Boolean =
                SubscriptionUtil.isSimHardwareVisible(context) &&
                    context.userManager.isAdminUser
        }
    }
}
